src/R/Components/Impl/Plots/Implementation/RPlotManager.cs

```diff
@@ -188,35 +283,143 @@ internal class RPlotManager : IRPlotManager {
+            } else if (plot.IsPlot) {
+                try {
+                    var img = plot.ToBitmapImage();
+                    await viewModel.PlotMessageLoadPlotAsync(plot.DeviceId, plot.PlotId, img, plot.DeviceNum, plot.ActivePlotIndex, plot.PlotCount);
+                    History.AddOrUpdate(viewModel.DeviceName, viewModel.DeviceId, plot.PlotId, viewModel.SessionProcessId, img);
+                } catch (Exception e) when (!e.IsCriticalException()) {
+                    await viewModel.PlotMessageLoadErrorAsync(plot.DeviceId, plot.PlotId, plot.DeviceNum, plot.ActivePlotIndex, plot.PlotCount);
+                }
+            } else if (plot.IsError) {
+                await viewModel.PlotMessageLoadErrorAsync(plot.DeviceId, plot.PlotId, plot.DeviceNum, plot.ActivePlotIndex, plot.PlotCount);
+            }
+        }
 
-        public Task ExportToPdfAsync(string outputFilePath) =>
-            ExportAsync(outputFilePath, _interactiveWorkflow.RSession.ExportToPdfAsync(outputFilePath, PixelsToInches(_lastPixelWidth), PixelsToInches(_lastPixelHeight)));
+        private async Task RemoveFromHistoryAsync(Guid plotId) {
+            await InteractiveWorkflow.Shell.SwitchToMainThreadAsync();
+            History.Remove(plotId);
+        }
 
-        public void EndLocatorMode() {
-            EndLocatorMode(LocatorResult.CreateNotClicked());
+        private async Task ClearHistoryAsync() {
+            await InteractiveWorkflow.Shell.SwitchToMainThreadAsync();
+            History.Clear();
         }
 
-        public void EndLocatorMode(LocatorResult result) {
-            var tcs = _locatorTcs;
-            _locatorTcs = null;
-            tcs?.SetResult(result);
-            _interactiveWorkflow.Shell.DispatchOnUIThread(() => SetLocatorModeUI(false));
-            LocatorModeChanged?.Invoke(this, null);
+        private async void RSession_Mutated(object sender, EventArgs e) {
+            await RSessionMutatedAsync();
         }
 
-        private void SetLocatorModeUI(bool locatorMode) {
-            if (VisualComponent != null) {
-                VisualComponent.SetLocatorMode(locatorMode);
+        private async Task RSessionMutatedAsync() {
+            try {
+                var deviceId = await InteractiveWorkflow.RSession.GetActivePlotDeviceAsync();
+                var deviceChanged = ActiveDeviceId != deviceId;
+                ActiveDeviceId = deviceId;
+
+                // Update all the devices in parallel
+                var visualComponents = _visualComponents.Values.ToArray();
+                var tasks = visualComponents.Select(v => v?.ViewModel.RefreshDeviceNameAsync());
+                await Task.WhenAll(tasks);
+
+                _interactiveWorkflow.ActiveWindow?.Container.UpdateCommandStatus(false);
+
+                if (deviceChanged) {
+                    ActiveDeviceChanged?.Invoke(this, EventArgs.Empty);
+                }
+
+            } catch (RException) {
```

Comment:
Not really. If method is called with `DoNotWait`, it will be swallowed automatically

---
In reply to: [76161851](https://github.com/Microsoft/RTVS/pull/2199#discussion_r76161851) [](ancestors = 76161851)